tests/integration/test_charm.py

```diff
@@ -0,0 +1,409 @@
+
+    await ops_test.model.applications[CONSUMER].set_config(consumer_config)
+    await ops_test.model.wait_for_idle(apps=[CONSUMER])
+
+    config_consumer = {"topic-name": topic_name, "extra-user-roles": "consumer"}
+    config_producer = {"topic-name": topic_name, "extra-user-roles": "producer"}
+
+    await asyncio.gather(
+        ops_test.model.deploy(
+            DATA_INTEGRATOR,
+            application_name=DATA_INTEGRATOR_CONSUMER,
+            channel="edge",
+            num_units=1,
+            series="jammy",
+            config=config_consumer,
+        ),
+        ops_test.model.deploy(
+            DATA_INTEGRATOR,
+            application_name=DATA_INTEGRATOR_PRODUCER,
+            channel="edge",
+            num_units=1,
+            series="jammy",
+            config=config_producer,
+        ),
+    )
+
+    await ops_test.model.wait_for_idle(apps=[DATA_INTEGRATOR_PRODUCER, DATA_INTEGRATOR_CONSUMER])
+
+    # test the active/waiting status for relation
+    await ops_test.model.wait_for_idle(apps=[DATA_INTEGRATOR_PRODUCER, DATA_INTEGRATOR_CONSUMER])
+    await ops_test.model.wait_for_idle(
+        apps=[KAFKA, DATA_INTEGRATOR_CONSUMER, DATA_INTEGRATOR_PRODUCER]
+    )
+
+    await ops_test.model.add_relation(KAFKA, DATA_INTEGRATOR_PRODUCER)
+    await ops_test.model.wait_for_idle(apps=[KAFKA, DATA_INTEGRATOR_PRODUCER])
+
+    await ops_test.model.add_relation(KAFKA, DATA_INTEGRATOR_CONSUMER)
+    await ops_test.model.wait_for_idle(apps=[KAFKA, DATA_INTEGRATOR_CONSUMER])
+
+    # get credential for Kafka
+    producer_credentials = await fetch_action_get_credentials(
+        ops_test.model.applications[DATA_INTEGRATOR_PRODUCER].units[0]
+    )
+
+    # get credential for Kafka
+    consumer_credentials = await fetch_action_get_credentials(
+        ops_test.model.applications[DATA_INTEGRATOR_CONSUMER].units[0]
+    )
+    logger.info(f"Producer credentials: {producer_credentials}")
+    logger.info(f"Consumer credentials: {consumer_credentials}")
+    # check for consumed messages in logs
+    res_producer = await fetch_action_start_process(
+        ops_test.model.applications[PRODUCER].units[0],
+        "start-process",
+        producer_credentials["kafka"]["endpoints"],
+        producer_credentials["kafka"]["username"],
+        producer_credentials["kafka"]["password"],
+        producer_credentials["kafka"]["topic"],
+    )
+
+    logger.info(f"Res: {res_producer}")
+
+    res_consumer = await fetch_action_start_process(
+        ops_test.model.applications[CONSUMER].units[0],
+        "start-process",
+        consumer_credentials["kafka"]["endpoints"],
+        consumer_credentials["kafka"]["username"],
+        consumer_credentials["kafka"]["password"],
+        consumer_credentials["kafka"]["topic"],
+        consumer_group_prefix=consumer_credentials["kafka"]["consumer-group-prefix"],
+    )
+
+    logger.info(f"Res: {res_consumer}")
+
+    time.sleep(20)
```

Comment:
**thought/non-blocking:** I wonder if `await asyncio.sleep(20)` is the right thing here. I know we sleep elsewhere in our tests, but part of me wonders if this does anything. Definitely not that important, but if you're re-running I'm curious what happens.

Reply:
I don't think it would matter. I mean the tests in this file are anyway sequential. `asyncio.sleep` is used when, while sleeping, you release the lock of the event loop, and another task (co-routing) can take it over. I have the feeling that here there would be only one task running anyway, so really, no other concurrent function would be making use of that waiting